---
title: "novel_PCA_projections"
author: "Prashit Parikh"
date: "2023-03-21"
output: html_document
---

The data directory necessary for analysis can be downloaded straight from biowulf here: https://hpc.nih.gov/~parikhpp/eyeIntegration_app_v2/data.tar.gz

```{r, include=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 20, fig.height = 10, fig.align = "center")

library(tidyverse)
library(plotly)
library(data.table)
library(matrixStats)
library(tximport)

# Working directory:
# The scripts folder within the EiaD_build repository (EiaD_build/scripts/)
```

In this analysis, we are going to attempt to project new PCA visualizations on an existing PCA space. This will be done using scaling and at the run_accession level, as that is the level at which our local data using different quantification methods was processed.

Loading in the necessary data files:

```{r, warning=FALSE}
# this file generated by scripts/pca_workup_data_prep.R

load('../data/EiaD_pca_analysis.Rdata')
emeta <- data.table::fread('../data/eyeIntegration22_meta_2023_03_03.csv.gz')

# This file is from OGVFB_BG/EiaD_snakerail/counts
gene_counts_matrix <- vroom::vroom('../data/gene_counts_matrix.csv.gz')
gene_counts_matrix <- gene_counts_matrix %>% column_to_rownames(var = "gene_id")
gene_counts_matrix <- gene_counts_matrix[, emeta$run_accession]

# http://duffel.rail.bio/recount3/human/new_annotations/gene_sums/human.gene_sums.G029.gtf.gz
gene_annotations <- 
  rtracklayer::import("../data/human.gene_sums.G029.gtf.gz") %>% 
  as.data.frame()

# http://duffel.rail.bio/recount3/human/new_annotations/gene_sums/human.exon_sums.G029.gtf.gz
exon_annotations <- 
  rtracklayer::import("../data/human.exon_sums.G029.gtf.gz") %>% 
  as.data.frame()
```

```{r, warning=FALSE}
#Functions for downstream analysis

do_pca <- function(gene_by_run, meta, ntop = 1000, scale){
  # count values for gene_by_sample
  # metadata rows must match columns of gene_by_run
  # quick checker for data mismatch
  if (!(colnames(gene_by_run) == meta$run_accession) %>% sum() == ncol(gene_by_run)){
    stop("Check metadata and matrix run name matching")
  }
  Pvars <- rowVars(log2(gene_by_run+1))
  select <- order(Pvars, decreasing = TRUE)[seq_len(min(ntop,
                                                        length(Pvars)))]
  if (scale){
    PCA <- prcomp(t(log2(gene_by_run[select, ]+1)), scale = TRUE, center = TRUE)
  } else {
    PCA <- prcomp(t(log2(gene_by_run[select, ]+1)), scale = FALSE, center = FALSE)
  }
  percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2),1)
  
  gene_names = PCA$rotation %>% rownames()
  dataGG = cbind(PCA$x, meta)
  out <- list(PCA, dataGG, percentVar, select, gene_names)
  out
}

dist_processor <- function(runs_as_row_matrix, metadata){
  # row.names must be given that match `run_accession` in metadata
  # matrix should be log2 (or whatever) normalized
  #euc_dist <- dist(t(log2(TPMall[select,]+1)), method = 'euclidean')
  if ((nrow(metadata) != nrow(runs_as_row_matrix))){
    stop("Matrix row num != metadata row num!")
  }
  if (metadata$run_accession != row.names(runs_as_row_matrix)){
    stop("Matrix row names != metadata row names!")
  }
  euc_dist <- dist(runs_as_row_matrix, method = 'euclidean')
  euc_dist <- as.matrix(euc_dist)

  # metadata:
  # metadata %>% select(run_accession, BioSample, study_accession, Cohort, Tissue,  Sub_Tissue, Source, Age) %>% unique()
  xy <- t(combn(colnames(euc_dist), 2))
  euc_dist_long <- data.frame(xy, dist=euc_dist[xy])
  colnames(euc_dist_long) <- c('run_accession', 'against', 'dist')
  euc_dist_long_meta <- euc_dist_long %>%
    left_join(metadata, by = 'run_accession')
  euc_dist_long_meta
}

```

```{r, warning=FALSE}
##################################################
# PCA time

# Keep only protein-coding genes
pc_ensgene <- gene_annotations %>% filter(gene_type == 'protein_coding') %>% pull(gene_id) %>% gsub('\\.\\d+$','',.)
pc_gene_counts_matrix <- gene_counts_matrix[pc_ensgene, emeta %>% filter(Source != "scRNA", Cohort == "Eye") %>% pull(run_accession)]

# all minus scRNA with only protein coding genes
out <- do_pca(pc_gene_counts_matrix %>% as.matrix(), emeta %>% filter(Source != "scRNA", Cohort == "Eye"), scale = TRUE)
PCA <- out[[1]]
dataGG <- out[[2]]
percentVar <- out[[3]]
select <- out[[4]]
gene_names <- out[[5]]
gene_names <- gene_names %>% str_extract(., "^.*(?=(\\.))")

# dist by PCA
pca_mat <- PCA$x[,1:2]
row.names(pca_mat) <- emeta %>% filter(Source != "scRNA", Cohort == "Eye") %>% pull(run_accession)
pca_dist <- dist_processor(pca_mat,
                           emeta %>% filter(Source != "scRNA", Cohort == "Eye"))
```

```{r, warning=FALSE}
# PCA Distance Metadata

pca_dist_metadata <-
  pca_dist %>%
  # metadata for the "against" sample
  left_join(emeta %>% dplyr::select(run_accession, 
                                    study_accession2 = study_accession, 
                                    Tissue2 = Tissue,
                                    Cohort2 = Cohort) %>% unique(), 
            by = c('against' = 'run_accession')) %>% 
  filter(study_accession != study_accession2, 
         study_accession != 'SRP273695', # incorrectly labeled samples
         run_accession != against,
         #study_accession2 != 'SRP273695',
         Cohort == 'Eye', 
         Cohort2 == 'Eye') %>% 
  as_tibble() %>% 
  group_by(run_accession, study_accession, Tissue, Sub_Tissue, Source, Age) %>% 
  slice_min(n = 20, order_by = dist) %>%
  ungroup() %>% 
  group_by(run_accession, study_accession, Tissue, Sub_Tissue, Source, Age, Tissue2) %>% 
  summarise(Count = n()) %>% 
  mutate(Tissue3 = paste0(Tissue2, ' (', Count,')')) %>% 
  summarise(Count = sum(Count[Tissue != Tissue2]),
            Tissue2 = paste(Tissue3, collapse = ',') )

# Plotting PCA output with distance values

eye_pca_data <- dataGG %>% filter(run_accession %in% pca_dist_metadata$run_accession) %>%
  left_join(pca_dist_metadata %>% ungroup() %>% 
              select(-study_accession, -Tissue, -Sub_Tissue, -Source, -Age), by = c("run_accession"))
```

This is what the PCA from our Snakrail data originally looks like:

```{r, warning=FALSE}
pcFirst <- 'PC1'
pcSecond <- 'PC2'

ggplotly(eye_pca_data %>%
  #filter(Source != 'scRNA') %>% 
  as_tibble() %>%
  ggplot(., aes(.data[[pcFirst]], .data[[pcSecond]])) +
  geom_point(size=3, aes(color=Tissue, shape = Source,
                         text = paste(study_accession, "\n", sample_accession, "\n", run_accession, "\n", Tissue, "\n",
                                      Sub_Tissue, "\n", Source, "\n", Age, "\n",
                                      Count, "\n", Tissue2))) +
  xlab(paste0(pcFirst, ": ",percentVar[str_extract(pcFirst, '\\d+') %>% as.integer()],"% variance")) +
  ylab(paste0(pcSecond, ": ",percentVar[str_extract(pcSecond, '\\d+') %>% as.integer()],"% variance")) + 
  cowplot::theme_cowplot() +
  scale_color_manual(values = c(pals::glasbey(), pals::alphabet2(), pals::alphabet2()) %>% unname()) +
  scale_fill_manual(values = c(pals::glasbey(), pals::alphabet2(), pals::alphabet2()) %>% unname()) +
  scale_shape_manual(values = 0:10), tooltip = 'text')
```

Now that we have a basic PCA plot which will serve as our PCA space, we can work on adding additional data layers to it! For this example to keep it brief, we will use kallisto samples I ran - RPE, Retina, and Cornea Tissue Types:

# Kallisto Data

```{r}
# Transcript to gene map
tx2gene <- exon_annotations %>% select(transcript_id, gene_id)

# Set of samples which were locally quantified
kallisto_runs <- c("DRR228774", "DRR228775", "DRR228776", "DRR228777", "DRR228778", "DRR228779", "DRR228780", "ERR5236630", "ERR5236631", "ERR5236632", "ERR5236633", "ERR5236634", "ERR5236635", "ERR5236636", "ERR5236637", "ERR5236638", "ERR5236639", "SRR11587762", "SRR11587763", "SRR11587764", "SRR11587765", "SRR11587766", "SRR11587767", "SRR11587768", "SRR11587769", "SRR11587770", "SRR11587771", "SRR11587772", "SRR11587773", "SRR11587774", "SRR11587775", "SRR11587776", "SRR11587777", "SRR11587778", "SRR14768214", "SRR14768215", "SRR14768216", "SRR14768217")

files <- file.path("../data/kallisto_multiple_tissue/", kallisto_runs, "abundance.tsv/abundance.tsv")

# Provide names to differentiate from original samples and from one another
names(files) <- c("kallisto_DRR228774", "kallisto_DRR228775", "kallisto_DRR228776", "kallisto_DRR228777", "kallisto_DRR228778", "kallisto_DRR228779", "kallisto_DRR228780", "kallisto_ERR5236630", "kallisto_ERR5236631", "kallisto_ERR5236632", "kallisto_ERR5236633", "kallisto_ERR5236634", "kallisto_ERR5236635", "kallisto_ERR5236636", "kallisto_ERR5236637", "kallisto_ERR5236638", "kallisto_ERR5236639", "kallisto_SRR11587762", "kallisto_SRR11587763", "kallisto_SRR11587764", "kallisto_SRR11587765", "kallisto_SRR11587766", "kallisto_SRR11587767", "kallisto_SRR11587768", "kallisto_SRR11587769", "kallisto_SRR11587770", "kallisto_SRR11587771", "kallisto_SRR11587772", "kallisto_SRR11587773", "kallisto_SRR11587774", "kallisto_SRR11587775", "kallisto_SRR11587776", "kallisto_SRR11587777", "kallisto_SRR11587778", "kallisto_SRR14768214", "kallisto_SRR14768215", "kallisto_SRR14768216", "kallisto_SRR14768217")

txi.kallisto.tsv <- tximport(files, type = "kallisto", tx2gene = tx2gene, ignoreAfterBar = TRUE)
head(txi.kallisto.tsv$counts)

# Only genes from PCA
kallisto_data <- txi.kallisto.tsv$counts %>% as.data.frame() %>% rownames_to_column(var="gene_id")
kallisto_data <- kallisto_data %>% mutate(gene_id = str_extract(gene_id, "^.*(?=(\\.))"))

kallisto_data <- kallisto_data %>% filter(gene_id %in% gene_names)

# Code chunk to insert missing columns

not_included <- gene_names[!gene_names %in% intersect(kallisto_data$gene_id, gene_names)]

for (i in not_included){
  
    data <- data.frame(gene_id = i, row.names = FALSE)
    kallisto_data <- bind_rows(kallisto_data, data)
}

# Replace NAs with 0
kallisto_data[is.na(kallisto_data)] <- 0

# Match order of gene_ids in kallisto data to columns from original PCA
kallisto_data <- kallisto_data[match(PCA$center %>% names() %>% str_extract(., "^.*(?=(\\.))"), kallisto_data$gene_id),] %>% remove_rownames(.) # remove row names so we can add new ones later
```

```{r}
# scale data prior to adding it to plot
kallisto_scaled_data <- scale(kallisto_data %>% column_to_rownames(var="gene_id") %>% 
                                as.matrix(), center=TRUE, scale = TRUE) %>% t()
```

# Testing the addition of new samples

```{r}
# Line up column names
names(PCA$center) <- kallisto_scaled_data %>% colnames()
names(PCA$scale) <- kallisto_scaled_data %>% colnames()

# project new data onto the PCA space
scaled_data <- scale(kallisto_scaled_data, PCA$center, PCA$scale) %*% PCA$rotation %>% as.data.frame()

# Add Metadata
scaled_data <- scaled_data %>% rownames_to_column(var="kallisto_accession")
scaled_data <- scaled_data %>% mutate(run_accession = str_extract(kallisto_accession, "[^_]+$"))

scaled_data <- scaled_data %>% left_join(emeta %>% filter(run_accession %in% kallisto_runs), by = "run_accession")
```

If everything scales as it's supposed to, then we can go and plot the data on top of our current PCA like this:

```{r, warning=FALSE}
ggplotly(eye_pca_data %>%
  #filter(Source != 'scRNA') %>% 
  as_tibble() %>%
  ggplot(., aes(.data[[pcFirst]], .data[[pcSecond]])) +
  geom_point(size=3, aes(color=Tissue, shape = Source,
                         text = paste(study_accession, "\n", sample_accession, "\n", run_accession, "\n", Tissue, "\n",
                                      Sub_Tissue, "\n", Source, "\n", Age, "\n",
                                      Count, "\n", Tissue2))) +
  xlab(paste0(pcFirst, ": ",percentVar[str_extract(pcFirst, '\\d+') %>% as.integer()],"% variance")) +
  ylab(paste0(pcSecond, ": ",percentVar[str_extract(pcSecond, '\\d+') %>% as.integer()],"% variance")) + 
  cowplot::theme_cowplot() +
  scale_color_manual(values = c(pals::glasbey(), pals::alphabet2(), pals::alphabet2()) %>% unname()) +
  scale_fill_manual(values = c(pals::glasbey(), pals::alphabet2(), pals::alphabet2()) %>% unname()) +
  scale_shape_manual(values = 0:10) +
    geom_point(size = 3, aes(PC1, PC2, text=paste(study_accession, "\n", sample_accession, "\n", kallisto_accession, "\n", Tissue, "\n",
                                      Sub_Tissue, "\n", Source, "\n", Age), color=Tissue, shape=Source), data=scaled_data), tooltip = 'text')
```