---
title: "linear_model"
author: "Prashit Parikh"
date: "2022-11-08"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
chooseCRANmirror(graphics=FLASE, ind=1)
knitr::opts_chunk$set(echo = TRUE)

#install.packages("tidyverse")
library(tidyverse)
#install.packages("caret")
library(caret)
#install.packages("lmvar")
library(lmvar)
#install.packages("tibble")
library(tibble)
#install.packages("ggiraph")
library(ggiraph)

# Set working directory
setwd("/Users/parikhpp/git/sex_inference")
```

**Running Prediction Model based on Gene Counts Data**

```{r}
### All data files necessary for analysis can be found here (files must be uncompressed prior to use):
# https://hpc.nih.gov/~parikhpp/sex_inference/insert_file_name_here.csv.gz

# Load data for genes of interest and labeled sex only
regression_data <- read.csv("data/inference_regression_data.csv") %>% 
  select(-recount_qc.aligned_reads..chrx, -recount_qc.aligned_reads..chry)

regression_data <- regression_data %>% mutate(sex_binary =
                                                case_when(
                                                  sex == "female" ~ 1,
                                                  sex == "male" ~ 0,))

# Split the data into an experiment and training set

set.seed(1020)
experiment.samples <- regression_data$sex %>% 
  createDataPartition(p = 0.5, list = FALSE)
training.data  <- regression_data[experiment.samples, ]
experiment.data <- regression_data[-experiment.samples, ]

# Split the training set into a train and test set

set.seed(1020)
training.samples <- training.data$sex %>% 
  createDataPartition(p = 0.5, list = FALSE)
train.data <- training.data[training.samples, ]
test.data <- training.data[-training.samples, ]

# Remove sex for modeling

train.data_model <- train.data %>% select(-sex, -run_accession)
test.data_model <- test.data %>% select(-sex, -run_accession)
experiment.data_model <- experiment.data %>% select(-sex, -run_accession)

model1 <- lm(sex_binary ~ .,data = train.data_model)
summary(model1)
model1$coefficients

# Trained Model
prediction <- predict(model1, newdata = train.data_model)
predicted.classes <- ifelse(prediction > 0.5, "female", "male")
# Model accuracy
observed.classes <- train.data$sex
mean(predicted.classes == observed.classes)

# Test our model
prediction2 <- predict(model1, newdata = test.data_model)
predicted.classes2 <- ifelse(prediction2 > 0.5, "female", "male")
# Model accuracy
observed.classes2 <- test.data$sex
mean(predicted.classes2 == observed.classes2)

# Experiment data
prediction3 <- predict(model1, newdata = experiment.data_model)
predicted.classes3 <- ifelse(prediction3 > 0.5, "female", "male")
# Model accuracy
observed.classes3 <- experiment.data$sex
mean(predicted.classes3 == observed.classes3)
```

Gene TPM regression model:

y = 0.817 + (0.001)·(XIST) + (-0.003)·(EIF1AY) + (0.001)·(KDM5D) + (0.003)·(UTY) + (-0.011)·(DDX3Y) + (-0.009)·(RPS4Y1)

**Running Prediction Model based on Alignment Data**

```{r, warning=FALSE}
# Load data for alignment to x and y chromosome only
regression_data_xy <- read.csv("data/inference_regression_data.csv") %>% 
  select(run_accession, recount_qc.aligned_reads..chrx, recount_qc.aligned_reads..chry, sex)

regression_data_xy <- regression_data_xy %>% mutate(sex_binary =
                                                case_when(
                                                  sex == "female" ~ 1,
                                                  sex == "male" ~ 0,))

# Split the data into an experiment and training set

set.seed(1020)
experiment.samples_xy <- regression_data_xy$sex %>% 
  createDataPartition(p = 0.5, list = FALSE)
training.data_xy  <- regression_data_xy[experiment.samples_xy, ]
experiment.data_xy <- regression_data_xy[-experiment.samples_xy, ]

# Split the training set into a train and test set

set.seed(1020)
training.samples_xy <- training.data_xy$sex %>% 
  createDataPartition(p = 0.5, list = FALSE)
train.data_xy <- training.data_xy[training.samples_xy, ]
test.data_xy <- training.data_xy[-training.samples_xy, ]

# Remove sex for modeling

train.data_model_xy <- train.data_xy %>% select(-sex, -run_accession)
test.data_model_xy <- test.data_xy %>% select(-sex, -run_accession)
experiment.data_model_xy <- experiment.data_xy %>% select(-sex, -run_accession)

model1 <- lm(sex_binary ~ ., data = train.data_model_xy, y = TRUE, x = TRUE)
summary(model1)
model1$coefficients

# Trained Model
prediction_xy <- predict(model1, newdata = train.data_model_xy)
predicted.classes_xy <- ifelse(prediction_xy > 0.5, "female", "male")
# Model accuracy
observed.classes_xy <- train.data_xy$sex
mean(predicted.classes_xy == observed.classes_xy)

# Test our model
prediction2_xy <- predict(model1, newdata = test.data_model_xy)
predicted.classes2_xy <- ifelse(prediction2_xy > 0.5, "female", "male")
# Model accuracy
observed.classes2_xy <- test.data_xy$sex
mean(predicted.classes2_xy == observed.classes2_xy)

# Experiment data
prediction3_xy <- predict(model1, newdata = experiment.data_model_xy)
predicted.classes3_xy <- ifelse(prediction3_xy > 0.5, "female", "male")
# Model accuracy
observed.classes3_xy <- experiment.data_xy$sex
mean(predicted.classes3_xy == observed.classes3_xy)
```

Alignment regression model:

y = 0.328 + (0.255)·(x_alignment) + (-10.296)·(y_alignment)

**Plotting our results across the gene counts and alignment based prediction data**

Train Data

```{r}
train_data_combined <- cbind(prediction, prediction_xy, predicted.classes,
                             predicted.classes_xy, observed.classes, observed.classes_xy, train.data$run_accession) %>% as.data.frame()

names(train_data_combined) <- c("prediction", "prediction_xy", "predicted.classes", "predicted.classes_xy", 
                                "observed.classes", "observed.classes_xy", "run_accession")

train_data_combined <- train_data_combined %>% mutate(correct = case_when(
                                                        predicted.classes == observed.classes ~ "Correct",
                                                        predicted.classes != observed.classes ~ "Incorrect",))
train_data_combined <- train_data_combined %>% mutate(correct_xy = case_when(
                                                        predicted.classes_xy == observed.classes_xy ~ "Correct",
                                                        predicted.classes_xy != observed.classes_xy ~ "Incorrect",))
train_data_combined <- train_data_combined %>% mutate(correct_samples = ifelse((correct == "Incorrect" | correct_xy == "Incorrect"),
                                                                               "Incorrectly Predicted", "Correctly Predicted"))

# Create Plot 
gg_point <- ggplot(data = train_data_combined) +
  geom_point_interactive(aes(x = prediction %>% as.numeric(), y = prediction_xy %>% as.numeric(),
    tooltip = run_accession, data_id = run_accession, color = correct_samples)) + theme_classic() +
  xlab("Gene Counts Based Prediction Probability") +
  ylab("Alignment Based Prediction Probability") +
  ggtitle("Gene Counts vs Alignment \n Sex Inference Probabilities")
girafe(ggobj = gg_point)
```

Test Data

```{r}
test_data_combined <- cbind(prediction2, prediction2_xy, predicted.classes2,
                             predicted.classes2_xy, observed.classes2, observed.classes2_xy, test.data$run_accession) %>% as.data.frame()

names(test_data_combined) <- c("prediction", "prediction_xy", "predicted.classes", "predicted.classes_xy", 
                                "observed.classes", "observed.classes_xy", "run_accession")

test_data_combined <- test_data_combined %>% mutate(correct = case_when(
                                                        predicted.classes == observed.classes ~ "Correct",
                                                        predicted.classes != observed.classes ~ "Incorrect",))
test_data_combined <- test_data_combined %>% mutate(correct_xy = case_when(
                                                        predicted.classes_xy == observed.classes_xy ~ "Correct",
                                                        predicted.classes_xy != observed.classes_xy ~ "Incorrect",))
test_data_combined <- test_data_combined %>% mutate(correct_samples = ifelse((correct == "Incorrect" | correct_xy == "Incorrect"),
                                                                               "Incorrectly Predicted", "Correctly Predicted"))

# Create Plot 
gg_point <- ggplot(data = test_data_combined) +
  geom_point_interactive(aes(x = prediction %>% as.numeric(), y = prediction_xy %>% as.numeric(),
    tooltip = run_accession, data_id = run_accession, color = correct_samples)) + theme_classic() +
  xlab("Gene Counts Based Prediction Probability") +
  ylab("Alignment Based Prediction Probability") +
  ggtitle("Gene Counts vs Alignment \n Sex Inference Probabilities")
girafe(ggobj = gg_point)
```

Experiment Data

```{r}
experiment_data_combined <- cbind(prediction3, prediction3_xy, predicted.classes3,
                             predicted.classes3_xy, observed.classes3, observed.classes3_xy, experiment.data$run_accession) %>% as.data.frame()

names(experiment_data_combined) <- c("prediction", "prediction_xy", "predicted.classes", "predicted.classes_xy", 
                                "observed.classes", "observed.classes_xy", "run_accession")

experiment_data_combined <- experiment_data_combined %>% mutate(correct = case_when(
                                                        predicted.classes == observed.classes ~ "Correct",
                                                        predicted.classes != observed.classes ~ "Incorrect",))
experiment_data_combined <- experiment_data_combined %>% mutate(correct_xy = case_when(
                                                        predicted.classes_xy == observed.classes_xy ~ "Correct",
                                                        predicted.classes_xy != observed.classes_xy ~ "Incorrect",))
experiment_data_combined <- experiment_data_combined %>% mutate(correct_samples = ifelse((correct == "Incorrect" | correct_xy == "Incorrect"),
                                                                               "Incorrectly Predicted", "Correctly Predicted"))

# Create Plot 
gg_point <- ggplot(data = experiment_data_combined) +
  geom_point_interactive(aes(x = prediction %>% as.numeric(), y = prediction_xy %>% as.numeric(),
    tooltip = run_accession, data_id = run_accession, color = correct_samples)) + theme_classic() +
  xlab("Gene Counts Based Prediction Probability") +
  ylab("Alignment Based Prediction Probability") +
  ggtitle("Gene Counts vs Alignment \n Sex Inference Probabilities")
girafe(ggobj = gg_point)
```

**Running Prediction Model based on Gene Counts Data with Cross-Validation**

```{r}
# Set folds for cross-validation
data_ctrl <- trainControl(method = "cv", number = 10)

# Train model
model_caret <- train(sex_binary ~ .,
                     data = train.data_model,                   
                     trControl = data_ctrl,
                     method = "lm")
model_caret$finalModel$coefficients

# Trained Model
prediction <- predict(model_caret, newdata = train.data_model)
predicted.classes <- ifelse(prediction > 0.5, "female", "male")
# Model accuracy
observed.classes <- train.data$sex
mean(predicted.classes == observed.classes)

# Test our model
prediction2 <- predict(model_caret, newdata = test.data_model)
predicted.classes2 <- ifelse(prediction2 > 0.5, "female", "male")
# Model accuracy
observed.classes2 <- test.data$sex
mean(predicted.classes2 == observed.classes2)

# Experiment data
prediction3 <- predict(model_caret, newdata = experiment.data_model)
predicted.classes3 <- ifelse(prediction3 > 0.5, "female", "male")
# Model accuracy
observed.classes3 <- experiment.data$sex
mean(predicted.classes3 == observed.classes3)
```

Gene TPM regression model:

y = 0.817 + (0.001)·(XIST) + (-0.003)·(EIF1AY) + (0.001)·(KDM5D) + (0.003)·(UTY) + (-0.011)·(DDX3Y) + (-0.009)·(RPS4Y1)

**Running Prediction Model based on Alignment Data with Cross-Validation**

```{r, warning=FALSE}
# Set folds for cross-validation
data_ctrl <- trainControl(method = "cv", number = 10)

# Train model
model_caret <- train(sex_binary ~ .,
                     data = train.data_model_xy,                   
                     trControl = data_ctrl,
                     method = "lm")                
model_caret$finalModel$coefficients



# Trained Model
prediction_xy <- predict(model_caret, newdata = train.data_model_xy)
predicted.classes_xy <- ifelse(prediction_xy > 0.5, "female", "male")
# Model accuracy
observed.classes_xy <- train.data_xy$sex
mean(predicted.classes_xy == observed.classes_xy)

# Test our model
prediction2_xy <- predict(model_caret, newdata = test.data_model_xy)
predicted.classes2_xy <- ifelse(prediction2_xy > 0.5, "female", "male")
# Model accuracy
observed.classes2_xy <- test.data_xy$sex
mean(predicted.classes2_xy == observed.classes2_xy)

# Experiment data
prediction3_xy <- predict(model_caret, newdata = experiment.data_model_xy)
predicted.classes3_xy <- ifelse(prediction3_xy > 0.5, "female", "male")
# Model accuracy
observed.classes3_xy <- experiment.data_xy$sex
mean(predicted.classes3_xy == observed.classes3_xy)
```

Alignment regression model:

y = 0.328 + (0.255)·(x_alignment) + (-10.296)·(y_alignment)

**Plotting our cross-validation results across the gene counts and alignment based prediction data**

Train Data

```{r}
train_data_combined <- cbind(prediction, prediction_xy, predicted.classes,
                             predicted.classes_xy, observed.classes, observed.classes_xy, train.data$run_accession) %>% as.data.frame()

names(train_data_combined) <- c("prediction", "prediction_xy", "predicted.classes", "predicted.classes_xy", 
                                "observed.classes", "observed.classes_xy", "run_accession")

train_data_combined <- train_data_combined %>% mutate(correct = case_when(
                                                        predicted.classes == observed.classes ~ "Correct",
                                                        predicted.classes != observed.classes ~ "Incorrect",))
train_data_combined <- train_data_combined %>% mutate(correct_xy = case_when(
                                                        predicted.classes_xy == observed.classes_xy ~ "Correct",
                                                        predicted.classes_xy != observed.classes_xy ~ "Incorrect",))
train_data_combined <- train_data_combined %>% mutate(correct_samples = ifelse((correct == "Incorrect" | correct_xy == "Incorrect"),
                                                                               "Incorrectly Predicted", "Correctly Predicted"))

# Create Plot 
gg_point <- ggplot(data = train_data_combined) +
  geom_point_interactive(aes(x = prediction %>% as.numeric(), y = prediction_xy %>% as.numeric(),
    tooltip = run_accession, data_id = run_accession, color = correct_samples)) + theme_classic() +
  xlab("Gene Counts Based Prediction Probability") +
  ylab("Alignment Based Prediction Probability") +
  ggtitle("Cross-Validated Gene Counts vs Alignment \n Sex Inference Probabilities")
girafe(ggobj = gg_point)
```

Test Data

```{r}
test_data_combined <- cbind(prediction2, prediction2_xy, predicted.classes2,
                             predicted.classes2_xy, observed.classes2, observed.classes2_xy, test.data$run_accession) %>% as.data.frame()

names(test_data_combined) <- c("prediction", "prediction_xy", "predicted.classes", "predicted.classes_xy", 
                                "observed.classes", "observed.classes_xy", "run_accession")

test_data_combined <- test_data_combined %>% mutate(correct = case_when(
                                                        predicted.classes == observed.classes ~ "Correct",
                                                        predicted.classes != observed.classes ~ "Incorrect",))
test_data_combined <- test_data_combined %>% mutate(correct_xy = case_when(
                                                        predicted.classes_xy == observed.classes_xy ~ "Correct",
                                                        predicted.classes_xy != observed.classes_xy ~ "Incorrect",))
test_data_combined <- test_data_combined %>% mutate(correct_samples = ifelse((correct == "Incorrect" | correct_xy == "Incorrect"),
                                                                               "Incorrectly Predicted", "Correctly Predicted"))

# Create Plot 
gg_point <- ggplot(data = test_data_combined) +
  geom_point_interactive(aes(x = prediction %>% as.numeric(), y = prediction_xy %>% as.numeric(),
    tooltip = run_accession, data_id = run_accession, color = correct_samples)) + theme_classic() +
  xlab("Gene Counts Based Prediction Probability") +
  ylab("Alignment Based Prediction Probability") +
  ggtitle("Cross-Validated Gene Counts vs Alignment \n Sex Inference Probabilities")
girafe(ggobj = gg_point)
```

Experiment Data

```{r}
experiment_data_combined <- cbind(prediction3, prediction3_xy, predicted.classes3,
                             predicted.classes3_xy, observed.classes3, observed.classes3_xy, experiment.data$run_accession) %>% as.data.frame()

names(experiment_data_combined) <- c("prediction", "prediction_xy", "predicted.classes", "predicted.classes_xy", 
                                "observed.classes", "observed.classes_xy", "run_accession")

experiment_data_combined <- experiment_data_combined %>% mutate(correct = case_when(
                                                        predicted.classes == observed.classes ~ "Correct",
                                                        predicted.classes != observed.classes ~ "Incorrect",))
experiment_data_combined <- experiment_data_combined %>% mutate(correct_xy = case_when(
                                                        predicted.classes_xy == observed.classes_xy ~ "Correct",
                                                        predicted.classes_xy != observed.classes_xy ~ "Incorrect",))
experiment_data_combined <- experiment_data_combined %>% mutate(correct_samples = ifelse((correct == "Incorrect" | correct_xy == "Incorrect"),
                                                                               "Incorrectly Predicted", "Correctly Predicted"))

# Create Plot 
gg_point <- ggplot(data = experiment_data_combined) +
  geom_point_interactive(aes(x = prediction %>% as.numeric(), y = prediction_xy %>% as.numeric(),
    tooltip = run_accession, data_id = run_accession, color = correct_samples)) + theme_classic() +
  xlab("Gene Counts Based Prediction Probability") +
  ylab("Alignment Based Prediction Probability") +
  ggtitle("Cross-Validated Gene Counts vs Alignment \n Sex Inference Probabilities")
girafe(ggobj = gg_point)
```

There is no noticeable difference in model performance when cross-validation is utilized.
