---
title: "Label Analysis"
output: 
  html_notebook:
    author: "David McGaughey"
    date: "2023-09-01"
    theme: flatly
    toc: true
    toc_float: true
---

# Introduction

The metamoRph_label.R script was used to "auto label" all of the conjunctiva,
cornea, lens, retina, rpe, and trabecular meshwork samples in the EiaD. Why not optic nerve? Because the others have at least two studies of samples.

The rough approach of metamoRph_label.R is to make make a subsample of the EiaD dataset (tissues listed above, with 10% randomly chosen tissues for each sample type (at least 5)) and:

1. run pca on the samples
2. use the PCs to train a lm / regression classifier for each tissue type
3. project the full dataset by multiplying the counts against the rotation matrix
4. use the trained model to label the projected dataset

Then do that 20 times. This "bootstrapping" approach is designed to not robustly *not* overfit the models.

The labels and scoring are imported here so I can hand assess the ML labels and identify problematic samples to remove.

The assessments are doing via running a PCA on all of the ocular samples (minus the AMD perturbed ones)
and looking for how the ML regression scores (a higher minimum_score means the ML model has lower confidence
in the call) and predicted labels vary across the dataset. I then hand assess each potential outlier in the context of their full metadata.


```{r}
library(tidyverse)
# remotes::install_github("davemcg/metamoRph")
library(metamoRph)
library(data.table)
# this file generated by scripts/pca_workup_data_prep.R
mat_all <- vroom::vroom("~/git/EiaD_build/counts/gene_counts.csv.gz") %>% data.table() %>% 
  filter(!grepl("ENST", Gene)) %>% 
  mutate(Gene = gsub(" \\(.*","",Gene)) #%>% 
mat_all <- mat_all[, lapply(.SD, sum, na.rm=TRUE), by='Gene' ]
genes <- mat_all$Gene
mat_all <- mat_all[,-1] %>% as.matrix()
row.names(mat_all) <- genes
mat_all[1:10,1:10]

emeta <- data.table::fread('../data/eyeIntegration23_meta_2023_09_01.csv.gz')

# from script/eigenProjecR_label.R
predictions <- read_tsv('../data/2023_09_01_ML_tissue_predictions.tsv.gz')
pred_values <- read_tsv('../data/2023_09e_01_ML_tissue_values.tsv.gz')

tissues <- c("Conjunctiva", "Cornea", "Lens", "Retina", "RPE", "Trabecular Meshwork")
```
# Stats
```{r, fig.width=6, fig.height=10}
emeta_filter <- emeta %>% 
  as_tibble() %>% 
  filter(study_accession != 'SRP012682') %>% 
  filter(Tissue %in% tissues) %>% 
  filter(!grepl("AMD", Perturbation)) %>% 
  select(sample_accession:study_title,Source_details ) %>% 
  unique()

emeta_filter %>% 
  mutate(Perturbation = case_when(grepl('MGS', Source_details) ~ Source_details),
         Perturbation = gsub("Adult Tissue","", Perturbation)) %>% 
  mutate(Sub_Tissue = case_when(is.na(Sub_Tissue) ~ '', TRUE ~ Sub_Tissue), 
         Source = case_when(is.na(Source) ~ '', TRUE ~ Source), 
         Age = case_when(is.na(Age) ~ '', TRUE ~ Age),
         Perturbation = case_when(is.na(Perturbation) ~ '', TRUE ~ Perturbation)) %>% 
  group_by(Tissue, Source, Sub_Tissue, Age, Perturbation, study_accession) %>% 
  summarise(`Study Count` = length(unique(study_accession)),
            `Sample Count` = n()) %>% 
  mutate(Tissue = case_when(grepl("Trab",Tissue) ~ 'TM', grepl("EyeLid", Tissue) ~ "Eye Lid", TRUE ~ Tissue),
         Source = case_when(grepl("Primary", Source) ~ "P. Culture",
                            TRUE ~ Source)
  ) %>% 
  ggplot(aes(x ='', y=`Sample Count`, fill = study_accession)) + 
  ggh4x::facet_nested(Tissue+Source+Sub_Tissue+Age ~ ., scale = 'free', space=  'free', switch = 'y',
                      strip = ggh4x::strip_nested(size = "variable"))+ 
  geom_bar(stat = 'identity') +
  xlab('') + 
  coord_flip()  + 
  theme_minimal() + 
  theme(strip.text.y.left = element_text(angle = 0, size = 12), text = element_text(size = 16))  + 
  theme(strip.background = element_rect(colour="gray", fill="gray"),
        strip.switch.pad.wrap = margin(t = 0, r = 0, b = 0, l = 0)) +
  theme(legend.position = "none") +
  scale_fill_manual(values = c(pals::alphabet() , pals::alphabet2(), pals::glasbey()) %>% unname())
```

# Top level look
## Run PCA 

Data built from scripts/pca_workup_data_prep.R and run on the non-AMD perturbed ocular bulk RNA-seq samples.

```{r, fig.height=7, fig.width=9}
library(metamoRph)
pca_eiad <- run_pca(mat_all[,(emeta_filter %>% pull(sample_accession))], emeta_filter)

PCA <- pca_eiad[[1]]
dataGG <- cbind(PCA$x, pca_eiad[[2]])
percentVar <- pca_eiad[[3]]

pcFirst <- 'PC1'
pcSecond <- 'PC2'
rotations <- c(pcFirst, pcSecond)

dataGG %>%
  #filter(Source != 'scRNA') %>% 
  as_tibble() %>% 
  mutate(Tissue = case_when(Tissue == 'Brain' ~ Tissue,
                            Cohort == 'Body' ~ 'Body',
                            TRUE ~ Tissue)) %>% 
  ggplot(., aes(.data[[pcFirst]], .data[[pcSecond]])) +
  geom_point(size=1, aes(color=Tissue, shape = Source)) +
  xlab(paste0(pcFirst, ": ",percentVar[str_extract(pcFirst, '\\d+') %>% as.integer()],"% variance")) +
  ylab(paste0(pcSecond, ": ",percentVar[str_extract(pcSecond, '\\d+') %>% as.integer()],"% variance")) + 
  cowplot::theme_cowplot() +
  scale_color_manual(values = c(pals::glasbey(), pals::alphabet2(), pals::alphabet2()) %>% unname()) +
  scale_fill_manual(values = c(pals::glasbey(), pals::alphabet2(), pals::alphabet2()) %>% unname()) +
  scale_shape_manual(values = 0:10) 
```

## Reactive plot

```{r, fig.width=10, fig.height=6}
library(plotly)
p <- dataGG %>%
  left_join(predictions, by = c("sample_accession" = "sample_id")) %>% 
  as_tibble() %>% 
  mutate(Tissue = case_when(Tissue == 'Brain' ~ Tissue,
                            Cohort == 'Body' ~ 'Body',
                            TRUE ~ Tissue),
         Mislabel = case_when(sample_label != predict ~ 'Yes', TRUE ~ 'No'),
         Label = paste(Mislabel, sample_accession, Sub_Tissue, Source, Age, sep = '\n')) %>%
  ggplot(., aes(.data[[pcFirst]], .data[[pcSecond]], label = Label)) +
  geom_point(size=1, aes(color=Tissue, shape = Source)) +
  xlab(paste0(pcFirst, ": ",percentVar[str_extract(pcFirst, '\\d+') %>% as.integer()],"% variance")) +
  ylab(paste0(pcSecond, ": ",percentVar[str_extract(pcSecond, '\\d+') %>% as.integer()],"% variance")) + 
  cowplot::theme_cowplot() +
  scale_color_manual(values = c(pals::glasbey(), pals::alphabet2(), pals::alphabet2()) %>% unname()) +
  scale_fill_manual(values = c(pals::glasbey(), pals::alphabet2(), pals::alphabet2()) %>% unname()) +
  scale_shape_manual(values = 0:10)
ggplotly(p)
```


# Split by Tissue and Color by Min Score
Where higher values (near 1) mean higher confidence in the label assignment

## PC1 and PC2
```{r, fig.width=12, fig.height=8}
pcFirst <- 'PC1'
pcSecond <- 'PC2'
dataGG %>%
  filter(Tissue %in% tissues) %>% 
  as_tibble() %>% 
  mutate(Tissue = case_when(Tissue == 'Brain' ~ Tissue,
                            Cohort == 'Body' ~ 'Body',
                            TRUE ~ Tissue)) %>% 
  left_join(pred_values, by = c('sample_accession' = 'sample_id')) %>% 
  mutate(mean_max_score = case_when(mean_max_score > 1 ~ 1, TRUE ~ mean_max_score)) %>% 
  ggplot(., aes(.data[[pcFirst]], .data[[pcSecond]])) +
  geom_point(size=3, aes(color=mean_max_score, shape = Source)) +
  xlab(paste0(pcFirst, ": ",percentVar[str_extract(pcFirst, '\\d+') %>% as.integer()],"% variance")) +
  ylab(paste0(pcSecond, ": ",percentVar[str_extract(pcSecond, '\\d+') %>% as.integer()],"% variance")) + 
  cowplot::theme_cowplot() +
  scale_color_viridis_c() +
  scale_shape_manual(values = 0:10) + facet_wrap(~Tissue)
```

## PC3 and PC4
```{r, fig.width=12, fig.height=8}
pcFirst <- 'PC3'
pcSecond <- 'PC4'
dataGG %>%
  filter(Tissue %in% tissues) %>% 
  as_tibble() %>% 
  mutate(Tissue = case_when(Tissue == 'Brain' ~ Tissue,
                            Cohort == 'Body' ~ 'Body',
                            TRUE ~ Tissue)) %>% 
  left_join(pred_values, by = c('sample_accession' = 'sample_id')) %>% 
  mutate(mean_max_score = case_when(mean_max_score > 1 ~ 1, TRUE ~ mean_max_score)) %>% 
  ggplot(., aes(.data[[pcFirst]], .data[[pcSecond]])) +
  geom_point(size=3, aes(color=mean_max_score, shape = Source)) +
  xlab(paste0(pcFirst, ": ",percentVar[str_extract(pcFirst, '\\d+') %>% as.integer()],"% variance")) +
  ylab(paste0(pcSecond, ": ",percentVar[str_extract(pcSecond, '\\d+') %>% as.integer()],"% variance")) + 
  cowplot::theme_cowplot() +
  scale_color_viridis_c() +
  scale_shape_manual(values = 0:10) + facet_wrap(~Tissue)
```

# Split by Tissue and Color by ML mislabel
Where lower values are more confident in the metamoRph tissue score

"Suspicious" samples are idenfied by three criteria:

1. Predicted label does not match true (?) label
2. The best ML score is above 0.5 (hand picked by looking at distribution of minimum ML score)
3. If the ensemble of incorrect predicted labels is more than half 
```{r, fig.width=12, fig.height=8}

sus <- predictions %>% filter(sample_label != predict | mean_max_score < 0.7 | miscall_count > 9)

for (i in c(1,3,5,7,9)){
  pcFirst <- paste0('PC', i)
  pcSecond <- paste0('PC', i+1)
  
  
  print(dataGG %>%
          filter(Tissue %in% tissues) %>% 
          as_tibble() %>% 
          mutate(Tissue = case_when(Tissue == 'Brain' ~ Tissue,
                                    Cohort == 'Body' ~ 'Body',
                                    TRUE ~ Tissue)) %>% 
          mutate(Suspicious = case_when(sample_accession %in% sus$sample_id ~ 'Yes')) %>% 
          ggplot(., aes(.data[[pcFirst]], .data[[pcSecond]])) +
          geom_point(size=3, aes(color=Suspicious, shape = Source)) +
          xlab(paste0(pcFirst, ": ",percentVar[str_extract(pcFirst, '\\d+') %>% 
                                                 as.integer()],"% variance")) +
          ylab(paste0(pcSecond, ": ",percentVar[str_extract(pcSecond, '\\d+') %>% 
                                                  as.integer()],"% variance")) + 
          cowplot::theme_cowplot() +
          scale_shape_manual(values = 0:10) + facet_wrap(~Tissue)
  )}

```

<!-- # UMAP -->
<!-- Add meta info for whether the metamoRph tissue labeling suspects and issue ("Sus") -->

<!-- ```{r} -->
<!-- custom.settings = umap::umap.defaults -->
<!-- custom.settings$n_neighbors = 30 -->
<!-- custom.settings$metric <- 'euclidean' -->
<!-- custom.settings$min_dist <- 0.5 -->
<!-- library(plotly) -->
<!-- umap <- umap::umap(PCA$x[,1:20], config = custom.settings) -->
<!-- ``` -->

<!-- ```{r, fig.width=15, fig.height=10} -->
<!-- p <- umap$layout %>% as_tibble(rownames = 'sample_accession') %>% -->
<!--   mutate(Suspicious = case_when(sample_accession %in% sus$sample_id ~ 'Yes')) %>% -->
<!--   left_join(dataGG, by = 'sample_accession') %>% -->
<!--   left_join(predictions, by = c("sample_accession" = "sample_id")) %>% -->
<!--   mutate(lab =  paste0(predict, ' (', mean_max_score, ')\n', sample_accession, '\n', Age)) %>% -->
<!--   # mutate(Tissue = case_when(Tissue == 'Brain' ~ Tissue, -->
<!--   #                          Cohort == 'Body' ~ 'Body', -->
<!--   #                          TRUE ~ Tissue)) %>% -->
<!--   ggplot(aes(x=V1,y=V2, label = lab)) + -->
<!--   geom_point(size=3, aes(color=Tissue, shape = Tissue)) + -->
<!--   cowplot::theme_cowplot() + -->
<!--   scale_color_manual(values = c(pals::glasbey(), pals::alphabet2(), pals::alphabet2()) %>% unname()) + -->
<!--   scale_fill_manual(values = c(pals::glasbey(), pals::alphabet2(), pals::alphabet2()) %>% unname()) + -->
<!--   scale_shape_manual(values = 0:50) -->
<!-- ggplotly(p) -->
<!-- ``` -->


<!-- ## Plot PCA and color by total reads -->
<!-- Was suspicious that PC2 was being driven by total counts ... nope! Not the case -->
<!-- ```{r, fig.width=12, fig.height=8} -->
<!-- run_meta <- read_csv('../salmon_counts/run_meta.csv.gz') -->
<!-- pcFirst <- 'PC1' -->
<!-- pcSecond <- 'PC2' -->
<!-- dataGG %>% -->
<!--   #filter(Source != 'scRNA') %>%  -->
<!--   as_tibble() %>%  -->
<!--   mutate(Tissue = case_when(Tissue == 'Brain' ~ Tissue, -->
<!--                             Cohort == 'Body' ~ 'Body', -->
<!--                             TRUE ~ Tissue)) %>%  -->
<!--   left_join( -->
<!--     run_meta %>% mutate(sample_accession = gsub('salmon_quant\\/|\\/aux_info\\/meta_info.json','',log))) %>%  -->
<!--   mutate(Sus = case_when(sample_accession %in% sus$sample_id ~ 'Yes') , -->
<!--          Age = case_when(is.na(Age) ~ 'None', TRUE ~ Age), -->
<!--          processed = log1p(processed) %>% round(., digits = 0)) %>%  -->
<!--   ggplot(., aes(.data[[pcFirst]], .data[[pcSecond]])) + -->
<!--   geom_point(size=1, aes(color=log1p(processed))) + -->
<!--   xlab(paste0(pcFirst, ": ",percentVar[str_extract(pcFirst, '\\d+') %>% as.integer()],"% variance")) + -->
<!--   ylab(paste0(pcSecond, ": ",percentVar[str_extract(pcSecond, '\\d+') %>% as.integer()],"% variance")) +  -->
<!--   cowplot::theme_cowplot() + -->
<!--   scale_color_viridis_c() + facet_wrap(~processed) -->
<!-- ``` -->

<!-- # Heatmap of metamoRph scoring -->
<!-- ```{r, fig.height=120, fig.width=10} -->
<!-- library(ComplexHeatmap) -->
<!-- hm_df <- left_join(predictions, pred_values %>% select(sample_id, Conjunctiva:`Trabecular Meshwork`)) %>% data.frame() -->
<!-- row.names(hm_df) <- hm_df$sample_id -->

<!-- hm_df <- hm_df %>%  -->
<!--   left_join(emeta %>%  -->
<!--               select(sample_accession, Sub_Tissue, Source, Age) %>% unique(),  -->
<!--             by = c('sample_id'='sample_accession'))  -->

<!-- ha_side = rowAnnotation(df = data.frame(ScientistLabel =  -->
<!--                                           hm_df$sample_label %>% factor(levels = unique(hm_df$sample_label)), -->
<!--                                         eigenLabel =  -->
<!--                                           hm_df$predict %>% factor(levels = unique(hm_df$sample_label)), -->
<!--                                         Source =  -->
<!--                                           hm_df$Source %>% factor(levels = unique(hm_df$Source))), -->
<!--                         col = list(ScientistLabel = c("Conjunctiva" = pals::alphabet2(20)[2] %>% unname(),  -->
<!--                                                       "Cornea" = pals::alphabet2(20)[3] %>% unname(), -->
<!--                                                       "Lens" = pals::alphabet2(20)[6] %>% unname(), -->
<!--                                                       "Retina" = pals::alphabet2(20)[10] %>% unname(), -->
<!--                                                       "RPE" = pals::alphabet2(20)[13] %>% unname(), -->
<!--                                                       "Trabecular Meshwork" = pals::alphabet2(20)[16] %>% unname()), -->
<!--                                    eigenLabel = c("Conjunctiva" = pals::alphabet2(20)[2] %>% unname(),  -->
<!--                                                   "Cornea" = pals::alphabet2(20)[3] %>% unname(), -->
<!--                                                   "Lens" = pals::alphabet2(20)[6] %>% unname(), -->
<!--                                                   "Retina" = pals::alphabet2(20)[10] %>% unname(), -->
<!--                                                   "RPE" = pals::alphabet2(20)[13] %>% unname(), -->
<!--                                                   "Trabecular Meshwork" = pals::alphabet2(20)[16] %>% unname()))) -->


<!-- Heatmap(hm_df[,8:13],  -->
<!--         right_annotation = ha_side, -->
<!--         cluster_rows = TRUE) -->
<!-- ``` -->
<!-- # Just the "mislabelled" ones -->

<!-- ```{r, fig.height=10, fig.width=5} -->

<!-- hm_df <- cbind(predictions, model_scores) -->
<!-- row.names(hm_df) <- hm_df$sample_id -->

<!-- hm_df <- hm_df %>%  -->
<!--   select(-predict) %>%  -->
<!--   left_join(calls %>% select(sample_id, predict)) %>%  -->
<!--   left_join(emeta %>%  -->
<!--               select(sample_accession, Sub_Tissue, Source, Age) %>% unique(),  -->
<!--             by = c('sample_id'='sample_accession')) %>%  -->
<!--   filter(sample_label != predict) -->

<!-- ha_side = rowAnnotation(df = data.frame(ScientistLabel =  -->
<!--                                           hm_df$sample_label %>% factor(levels = unique(hm_df$sample_label)), -->
<!--                                         eigenLabel =  -->
<!--                                           hm_df$predict %>% factor(levels = unique(hm_df$sample_label)), -->
<!--                                         Source =  -->
<!--                                           hm_df$Source %>% factor(levels = unique(hm_df$Source))), -->
<!--                         col = list(ScientistLabel = c("Conjunctiva" = pals::alphabet2(20)[2] %>% unname(),  -->
<!--                                                       "Cornea" = pals::alphabet2(20)[3] %>% unname(), -->
<!--                                                       "Lens" = pals::alphabet2(20)[6] %>% unname(), -->
<!--                                                       "Retina" = pals::alphabet2(20)[10] %>% unname(), -->
<!--                                                       "RPE" = pals::alphabet2(20)[13] %>% unname(), -->
<!--                                                       "Trabecular Meshwork" = pals::alphabet2(20)[16] %>% unname()), -->
<!--                                    eigenLabel = c("Conjunctiva" = pals::alphabet2(20)[2] %>% unname(),  -->
<!--                                                   "Cornea" = pals::alphabet2(20)[3] %>% unname(), -->
<!--                                                   "Lens" = pals::alphabet2(20)[6] %>% unname(), -->
<!--                                                   "Retina" = pals::alphabet2(20)[10] %>% unname(), -->
<!--                                                   "RPE" = pals::alphabet2(20)[13] %>% unname(), -->
<!--                                                   "Trabecular Meshwork" = pals::alphabet2(20)[16] %>% unname()))) -->


<!-- Heatmap(hm_df[,6:8],  -->
<!--         right_annotation = ha_side, cluster_rows = FALSE) -->
<!-- ``` -->

# One plot per study with any potential mis labels

```{r, fig.height=10, fig.width=10}
library(ComplexHeatmap)
studies_with_sus <- emeta %>% filter(sample_accession %in% sus$sample_id) %>% pull(study_accession) %>% unique()

full_df <- cbind(predictions, pred_values %>% select(Conjunctiva:`Trabecular.Meshwork`)) 
for (i in studies_with_sus){
  #print(i)
  hm_df <- left_join(predictions, pred_values %>% select(sample_id, Conjunctiva:`Trabecular.Meshwork`) %>% unique(), by = 'sample_id') %>% 
    left_join(emeta %>% select(sample_id = sample_accession, study_accession, Sub_Tissue, Source, Age), by ='sample_id') %>% 
    filter(study_accession %in% i) %>% 
    data.frame() %>% unique()
  row.names(hm_df) <- hm_df$sample_id
  
  
  ha_side = rowAnnotation(df = data.frame(ScientistLabel = 
                                            hm_df$sample_label %>% factor(levels = unique(full_df$sample_label)),
                                          eigenLabel = 
                                            hm_df$predict %>% factor(levels = unique(full_df$sample_label)),
                                          Source = 
                                            hm_df$Source %>% factor(levels = unique(hm_df$Source)),
                                          Age = 
                                            hm_df$Age %>% factor(levels = unique(hm_df$Age))),
                          col = list(ScientistLabel = c("Conjunctiva" = pals::alphabet2(20)[2] %>% unname(), 
                                                        "Cornea" = pals::alphabet2(20)[3] %>% unname(),
                                                        "Lens" = pals::alphabet2(20)[6] %>% unname(),
                                                        "Retina" = pals::alphabet2(20)[10] %>% unname(),
                                                        "RPE" = pals::alphabet2(20)[13] %>% unname(),
                                                        "Trabecular Meshwork" = pals::alphabet2(20)[16] %>% unname()),
                                     eigenLabel = c("Conjunctiva" = pals::alphabet2(20)[2] %>% unname(), 
                                                    "Cornea" = pals::alphabet2(20)[3] %>% unname(),
                                                    "Lens" = pals::alphabet2(20)[6] %>% unname(),
                                                    "Retina" = pals::alphabet2(20)[10] %>% unname(),
                                                    "RPE" = pals::alphabet2(20)[13] %>% unname(),
                                                    "Trabecular Meshwork" = pals::alphabet2(20)[16] %>% unname())))
  
  
  print(Heatmap(as.matrix(hm_df[,8:13]),  col=circlize::colorRamp2(c(0, 0.5, 1), c("blue", "white", "red")),
                right_annotation = ha_side, cluster_rows = FALSE, column_title = i))
  
  pcFirst <- "PC1"
  pcSecond <- 'PC2'
  print(dataGG %>%
          filter(study_accession == i, Tissue %in% tissues) %>% 
          as_tibble() %>% 
          mutate(Tissue = case_when(Tissue == 'Brain' ~ Tissue,
                                    Cohort == 'Body' ~ 'Body',
                                    TRUE ~ Tissue)) %>% 
          mutate(Suspicious = case_when(sample_accession %in% sus$sample_id ~ 'Yes', TRUE ~ 'No')) %>% 
          ggplot(., aes(.data[[pcFirst]], .data[[pcSecond]], label = sample_accession)) +
          geom_point(data = dataGG %>% 
                       filter(study_accession != i) %>% 
                       filter(Tissue %in% tissues),
                     size=2, 
                     aes(x =.data[[pcFirst]], y= .data[[pcSecond]]), color = 'grey' ) +
          geom_point(size=3, aes(color=Suspicious, shape = Source)) +
          
          ggrepel::geom_label_repel( size = 2.5, max.overlaps = Inf) +
          xlab(paste0(pcFirst, ": ",percentVar[str_extract(pcFirst, '\\d+') %>% 
                                                 as.integer()],"% variance")) +
          ylab(paste0(pcSecond, ": ",percentVar[str_extract(pcSecond, '\\d+') %>% 
                                                  as.integer()],"% variance")) + 
          cowplot::theme_cowplot() +
          scale_shape_manual(values = 0:10) + facet_wrap(~Tissue, ncol = 2)
  )
  
  pcFirst <- "PC3"
  pcSecond <- 'PC4'
  print(dataGG %>%
          filter(study_accession == i,
                 Tissue %in% tissues) %>% 
          as_tibble() %>% 
          mutate(Tissue = case_when(Tissue == 'Brain' ~ Tissue,
                                    Cohort == 'Body' ~ 'Body',
                                    TRUE ~ Tissue)) %>% 
          mutate(Suspicious = case_when(sample_accession %in% sus$sample_id ~ 'Yes', TRUE ~ 'No')) %>% 
          ggplot(., aes(.data[[pcFirst]], .data[[pcSecond]], label = sample_accession)) +
          geom_point(data = dataGG %>% 
                       filter(study_accession != i) %>% 
                       filter(Tissue %in% tissues), 
                     size=2, 
                     aes(x =.data[[pcFirst]], y= .data[[pcSecond]]), color = 'grey' ) +
          geom_point(size=3, aes(color=Suspicious, shape = Source)) +
          ggrepel::geom_label_repel(size = 2.5, max.overlaps = Inf) +
          xlab(paste0(pcFirst, ": ",percentVar[str_extract(pcFirst, '\\d+') %>% 
                                                 as.integer()],"% variance")) +
          ylab(paste0(pcSecond, ": ",percentVar[str_extract(pcSecond, '\\d+') %>% 
                                                  as.integer()],"% variance")) + 
          cowplot::theme_cowplot() +
          scale_shape_manual(values = 0:10) + facet_wrap(~Tissue, ncol = 2)
  )
}
```



# Hand picked samples to remove

Picked by using the plots above and checking out the full metadata to see if there were any other explanations for what appear to be outliers (an example is SRP105756 which has many early fetal retina tissues or SRP097696 which has a surprising flow sorted retina which only keeps endothelial cells).
```{r}
eigen_hand_removed <- c('SRS6509684',
                        'SRS390609',
                        'SRS542573','SRS542574',
                        'SRS1478834', #
                        'SRS8145606', #
                        'SRS360123','SRS360124', #
                        'SRS7504868','SRS7504869', #
                        'SRS523835','SRS523813',
                        'SRS1955494')

predictions %>% left_join(emeta %>% select(study_accession, sample_accession, Sub_Tissue, Source, Age), by = c('sample_id'='sample_accession')) %>% as_tibble() %>% filter(sample_id %in% eigen_hand_removed) %>% arrange(sample_label)

predictions %>% 
  left_join(emeta %>% ungroup() %>%  dplyr::select(study_accession, sample_accession, Sub_Tissue, Source, Age), 
            by = c('sample_id'='sample_accession')) %>% 
  as_tibble() %>% 
  filter(sample_id %in% eigen_hand_removed) %>% 
  arrange(sample_label) %>% group_by(sample_label, Sub_Tissue, Source, Age) %>% summarise(Count = n(), samples = paste0(sample_id, collapse = ', '))
```

```{r, fig.width=10, fig.height=7}
for (i in c(1,3,5,7)){
  pcFirst <- paste0('PC', i)
  pcSecond <- paste0('PC', i+1)
  print(dataGG %>%
          as_tibble() %>% 
          mutate(Tissue = case_when(Tissue == 'Brain' ~ Tissue,
                                    Cohort == 'Body' ~ 'Body',
                                    TRUE ~ Tissue)) %>% 
          mutate(Removed = case_when(sample_accession %in% eigen_hand_removed ~ 'Yes')) %>% 
          ggplot(., aes(.data[[pcFirst]], .data[[pcSecond]])) +
          geom_point(size=1, aes(color=Removed, shape = Source)) +
          xlab(paste0(pcFirst, ": ",percentVar[str_extract(pcFirst, '\\d+') %>% as.integer()],"% variance")) +
          ylab(paste0(pcSecond, ": ",percentVar[str_extract(pcSecond, '\\d+') %>% as.integer()],"% variance")) + 
          cowplot::theme_cowplot() +
          scale_color_manual(values = c(pals::glasbey(), pals::alphabet2(), pals::alphabet2()) %>% unname()) +
          scale_fill_manual(values = c(pals::glasbey(), pals::alphabet2(), pals::alphabet2()) %>% unname()) +
          scale_shape_manual(values = 0:10) +
          facet_wrap(~Tissue)
  )
}

```

# Output hand removed IDs
```{r}
write(eigen_hand_removed, file = '../data/excluded_samples.txt')
```

```{r}
devtools::session_info()
```



